### PR TITLE
Added qlty chopping features and DVC Live

### PR DESCRIPTION
**Pull request by TibbersHao**
This PR covers 2 major feature upgrades:

1. Added codes for `qlty` chopping for each frame inside `tiled_dataset`, and a corresponding `collate` function to accompany the patch number when passing into `DataLoader`.
2. Added `DVC Live` codes inside `train_segmentation()` function to save loss and metrics under the unique saving directory, along with the `model.pt`.

Note:
As @xiaoyachong mentioned, upon first time testing, git config might be needed for dvc to function properly. This does not auto commit losses to GitHub repo.
```
git config --global [user.name](http://user.name/) "John Doe"

git config --global user.email "johndoe@email.com"
```

### CODE REVIEW

File: example_yamls/example_tunet.yaml [hunk 1]
```diff
@@ -1,30 +1,31 @@
...
   criterion: "CrossEntropyLoss"
-  weights: null
+  weights: [1, 2, 4]
   learning_rate: 0.1
   activation: "ReLU"
   normalization: "BatchNorm2d"
   convolution: "Conv2d"
 
+  qlty_window: 50
+  qlty_step: 30
```
Comment (taxe10) [inline example_yamls/example_tunet.yaml:26]: As we discussed, the default value for the qlty window is too small for a unet of depth 4. Please modify accordingly.

**Review by taxe10 — CHANGES_REQUESTED** [review]
PR looks good. I left some minor comments.
Additionally, I was getting an error from `dvc` due to the init step, as shown below:
```
ERROR: failed to initiate DVC - /Users/user/mlex_dlsia_segmentation_prototype/.dvc is ignored by your SCM tool. 
Make sure that it's tracked, for example, by adding '!.dvc' to .gitignore.
```
Could you please check in with @xiaoyachong and update the init steps in the README file?

File: src/seg_utils.py [hunk 2]
```diff
@@ -2,8 +2,30 @@
...
+        data, mask = zip(*batch)
+        concated_data = torch.cat(data, dim=0) # concat on the first dim without introducing another dim -> keep in the 4d realm
+        concated_mask = torch.cat(mask, dim=0)
+        #print(f'concated_data shape: {concated_data.shape}')
+        #print(f'concated_mask shape: {concated_mask.shape}')
+        return concated_data, concated_mask
+    elif isinstance(elem, torch.Tensor) and elem.ndim == 4:
+        #print(f'batch size: {len(batch)}')
+        concated_data = torch.cat(batch, dim=0) # concat on the first dim without introducing another dim -> keep in the 4d realm
+        #print(f'concated_data shape: {concated_data.shape}')
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/seg_utils.py:25]: Done

File: src/seg_utils.py [hunk 3]
```diff
@@ -22,20 +44,22 @@ def train_val_split(dataset, parameters):
...
+        train_loader = DataLoader(train_data, **train_loader_params, collate_fn=custom_collate)
+        val_loader = DataLoader(val_data, **val_loader_params, collate_fn=custom_collate)
     else:
         train_size = len(dataset) - val_size
         train_data, val_data = random_split(dataset, [train_size, val_size])
-        train_loader = DataLoader(train_data, **train_loader_params)
-        val_loader = DataLoader(val_data, **val_loader_params)
-
+        train_loader = DataLoader(train_data, **train_loader_params, collate_fn=custom_collate)
+        val_loader = DataLoader(val_data, **val_loader_params, collate_fn=custom_collate)
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/seg_utils.py:62]: Good suggestion, implemented.

File: src/seg_utils.py [hunk 4]
```diff
@@ -129,171 +153,181 @@ def train_segmentation(
...
+                running_train_loss += loss.item()
+            if scheduler is not None:
+                scheduler.step()
+
+            # compute validation step
+            if validationloader is not None:
+                with torch.no_grad():
+                    for x, y in validationloader:
+                        x = x.type(torch.FloatTensor)
                         y = y.type(torch.LongTensor)
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/seg_utils.py:226]: Not from tiled. I got some `ByteType` tensors after introducing `qlty` cropping, still not sure why this happened in some cases but this is a quick solution to let the pipeline running.

File: src/seg_utils.py [hunk 5]
```diff
@@ -129,171 +153,181 @@ def train_segmentation(
...
+                F1_validation_trace_micro.append(F1_val_micro)
+                F1_validation_trace_macro.append(F1_val_macro)
+
+                live.log_metric("train/loss", loss)
+                live.log_metric("train/F1_micro", F1_micro)
+                live.log_metric("train/F1_macro", F1_macro)
+                live.log_metric("val/loss", val_loss)
+                live.log_metric("val/F1_micro", F1_val_micro)
+                live.log_metric("val/F1_macro", F1_val_macro)
+                live.next_step()
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/seg_utils.py:272]: I believe this will be discussed and covered in the other PR.

File: src/seg_utils.py [hunk 6]
```diff
@@ -313,21 +347,21 @@ def train_segmentation(
...
 
 # Segmentation
-def segment(net, device, inference_loader):
+def segment(net, device, inference_loader, qlty_object):
     net.to(device)   # send network to GPU
-    seg = None
-    for batch in inference_loader:
+    patch_preds = [] # store results for patches
+    for idx, batch in enumerate(inference_loader):
+        print(f'{idx+1}/{len(inference_loader)}')
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/seg_utils.py:354]: This is helpful to visualize the progress of frame by frame segmentation, and I used that as diagnostics for the saving issue that I am working on.

File: src/utils.py [hunk 7]
```diff
@@ -12,30 +12,24 @@ def create_directory(path):
...
-        else:
-            last_container = last_container[key]
+    last_container = last_container.create_container(key=uid)
 
     metadata={
-        'data_tiled_uri': data_tiled_uri,
-        'mask_uri': mask_tiled_uri, 
+        'data_tiled_uri': tiled_dataset.data_tiled_uri,
+        'mask_uri': tiled_dataset.mask_tiled_uri, 
+        'mask_idx': tiled_dataset.mask_idx,
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/utils.py:28]: As @Wiebke requested for the Inference-Preview mode, we would like to also save the `mask_idx` as part of the result metadata.

File: src/utils.py [hunk 8]
```diff
@@ -12,30 +12,24 @@ def create_directory(path):
...
         'uid': uid,
         'model': model, 
         }
+    print(f'metadata: {metadata}')
     seg_result = last_container.write_array(key="seg_result", array=seg_result, metadata=metadata)
     print("Segmentaion result array saved in following uri: ", seg_result.uri)
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/utils.py:34]: This print should be out, no need after I finished testing. Done.

File: src/tiled_dataset.py [hunk 9]
```diff
@@ -1,51 +1,90 @@
...
+                data = torch.from_numpy(data).unsqueeze(0).unsqueeze(0)
+                data_patches = self.qlty_object.unstitch(data)
+                return data_patches
+            
             else:
-                return data
\ No newline at end of file
+                data = self.data_client[idx,]
+                # Change to 4d array for qlty requirement
+                data = torch.from_numpy(data).unsqueeze(0).unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/tiled_dataset.py:88]: Marked!

File: src/parameters.py [hunk 10]
```diff
@@ -1,20 +1,35 @@
...
+
 #===========================================General Parameters===========================================#
 class TrainingParameters(BaseModel):
     network: str = Field(description="type of dlsia network used")
     num_classes: int = Field(description="number of classes as output channel")
     num_epochs: int = Field(default=10, description="number of epochs")
     optimizer: str = Field(default="Adam", description="optimizer used for training")
     criterion: str = Field(default="CrossEntropyLoss", description="criterion for loss")
-    weights: Optional[str] = Field(default=None, description="weights per class for imbalanced labeling")
+    weights: Optional[List[int]] = Field(default=None, description="weights per class for imbalanced labeling")
```
Comment (Wiebke) [inline src/parameters.py:21]: I believe we settled on leaving this as a string in the yaml file, but parsing it to a list of floats (unless PyTorch expects integers).

A reasonable alternative may also be to use weights according to the class distribution and make this a boolean parameter. The class imbalance could be computed during mask exporting and become additional meta data.

File: example_yamls/example_msdnet_customdil.yaml [hunk 11]
```diff
@@ -1,30 +1,31 @@
...
+  uid: uid0020
 
 model_parameters:
   network: "MSDNet"
   num_classes: 3
   num_epochs: 3
   optimizer: "Adam"
   criterion: "CrossEntropyLoss"
-  weights: null
+  weights: [1, 2, 4]
```
Comment (Wiebke) [inline example_yamls/example_msdnet_customdil.yaml:19]: Not a string of floats `[0.9, 0.1, 0.0]`?

File: example_yamls/example_msdnet_maxdil.yaml [hunk 12]
```diff
@@ -1,34 +1,35 @@
...
+  qlty_step: 30
+  qlty_border: 3
+  
   shuffle_train: True
-  batch_size_train: 1
+  batch_size_train: 3
   shuffle_val: True
-  batch_size_val: 1
+  batch_size_val: 2
   shuffle_inference: False
```
Comment (Wiebke) [inline example_yamls/example_msdnet_maxdil.yaml:33]: Should we also remove `shuffle_inference` and `shuffle_validation` already?

File: notebooks/smsnet_ensemble.ipynb [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+    "dataset.mask_client"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def custom_collate(batch):\n",
```
Comment (Wiebke) [inline notebooks/smsnet_ensemble.ipynb:127]: Is this function different from the one defined in `seg_utils.py`?

File: notebooks/smsnet_ensemble.ipynb [hunk 14]
```diff
@@ -0,0 +1,493 @@
...
+    "    list_of_smsnet.append(SMSNetwork_from_file(network))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def segment(net, device, inference_loader, qlty_object):\n",
```
Comment (Wiebke) [inline notebooks/smsnet_ensemble.ipynb:397]: Same here, this function appears to be defined elsewhere already and could be imported.

File: src/tiled_dataset.py [hunk 15]
```diff
@@ -1,51 +1,90 @@
...
         '''
+        self.data_tiled_uri = data_tiled_uri
         self.data_client = from_uri(data_tiled_uri, api_key=data_tiled_api_key)
+        self.mask_tiled_uri = mask_tiled_uri
         if mask_tiled_uri:
-            self.mask_client = from_uri(mask_tiled_uri, api_key=mask_tiled_api_key)
+            self.mask_client_one_up = from_uri(mask_tiled_uri, api_key=mask_tiled_api_key)
+            self.mask_client = self.mask_client_one_up['mask']
+            self.mask_idx = [int(idx) for idx in self.mask_client_one_up.metadata['mask_idx']]
         else:
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/tiled_dataset.py:40]: Added an `assert` in `train.py`, done.

File: src/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Wiebke) [reply src/utils.py:28]: That is a good point, Tanny.
I thought it would be a good idea to redundantly include these here to be able to directly map segmentation result to original data, but this can indeed also be retrieved through the mask uri. 

On a different note, if `mask_tiled_uri` is not given (in the segmental all case), does Tiled give us errors when one of the meta data values is `None`?

**Review by Wiebke — COMMENTED** [review]
Looks good. I left some questions, mostly regarding the `weights` parameter and metadata when writing back to Tiled. 

My comments on the notebook may be not relevant if this is just in preparation for integrating this as a new model.

File: src/parameters.py (hunk 10 above)
Comment (TibbersHao) [reply src/parameters.py:21]: I thought the example weights I got was in the format of ` [1.0, 1.0, 1.0]`, double checking on this, would this be a string or essentially this is a list in yaml file.

I have set this to a str earlier, and right now this is what I have `weights: Optional[List[float]]` in parameter validation. Will check again tomorrow.

File: example_yamls/example_msdnet_customdil.yaml (hunk 11 above)
Comment (TibbersHao) [reply example_yamls/example_msdnet_customdil.yaml:19]: The weights could also take `None` (null in yaml) if we don't want to assign weights for classes, this is just an example. 

I've added an array for the incoming commit.

File: example_yamls/example_msdnet_maxdil.yaml (hunk 12 above)
Comment (TibbersHao) [reply example_yamls/example_msdnet_maxdil.yaml:33]: Both have already been taken out in the next commit, done.

File: notebooks/smsnet_ensemble.ipynb (hunk 13 above)
Comment (TibbersHao) [reply notebooks/smsnet_ensemble.ipynb:127]: Most of the SMSNet stuff is similar to other models. 

FYI the implementation of SMSNet Ensemble has been ingested into src scripts. You can safely refer to those and discard the notebook.

File: notebooks/smsnet_ensemble.ipynb (hunk 14 above)
Comment (TibbersHao) [reply notebooks/smsnet_ensemble.ipynb:397]: Same as above.

File: src/seg_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (xiaoyachong) [reply src/seg_utils.py:258]: > Is the DVC report (html) updated every epoch? I think we can use `live.make_report()` for this, so that the loss plot is updated live. Maybe this belongs to #11 @xiaoyachong

Yes, html is updated every epoch. 'live.next_step()' will automatically call 'live.make_report()'.

**Comment by xiaoyachong** [comment]
> PR looks good. I left some minor comments. Additionally, I was getting an error from `dvc` due to the init step, as shown below:
> 
> ```
> ERROR: failed to initiate DVC - /Users/user/mlex_dlsia_segmentation_prototype/.dvc is ignored by your SCM tool. 
> Make sure that it's tracked, for example, by adding '!.dvc' to .gitignore.
> ```
> 
> Could you please check in with @xiaoyachong and update the init steps in the README file?

Maybe you can try to run 'dvc init' in the repository before training the model.

**Comment by TibbersHao** [comment]
I forgot to leave a comment last night, new commits ready to be reviewed. 

What has been covered:
1. Changes mentioned in comments above
2. SMSNet Ensemble method has been ingested into the source code
